Refactor ActiveRecord Signed ID to use global `Rails.application.message_verifiers`

AliSepehri opened the pull request:
### Ref: #54357 

This change ensures a unified configuration for all message verifiers, making it easier to rotate secrets and upgrade signing algorithms. See [message_verifiers](https://api.rubyonrails.org/classes/Rails/Application.html#method-i-message_verifiers) for more details.

### Details  
As suggested by @jonathanhefner [here](https://github.com/rails/rails/pull/54371/files#r1931046047), the best approach to resolve **issue #54357** is to use **MessageVerifiers**, similar to how `generated_token_verifier` was implemented previously.  

`config.active_record.use_legacy_signed_id_verifier` provides a smooth transition to this goal. It can be set to:

- `:generate_and_verify` (default) - Generate and verify signed IDs using the following legacy options:

```ruby
  { digest: "SHA256", serializer: JSON, url_safe: true }
```

- `:verify` - Generate and verify signed IDs using options from `Rails.application.message_verifiers`, but fall back to verifying with the same options as `:generate_and_verify`.

- `false` - Generate and verify signed IDs using options from `Rails.application.message_verifiers` only.

> **WARNING:** If `ActiveRecord::Base.signed_id_verifier_secret` is set in your application, Signed ID will not use global `message_verifiers`. 

---
These changes will enable proper usage of `rotate` and `on_rotation` as shown in the example below:  

```ruby
# Rotate verifier for Signed ID
Rails.application.message_verifiers.rotate do |salt|
  next nil if salt != 'active_record/signed_id'

  secret_generator = ->(_) { 'old_secret' }
  { secret_generator: secret_generator, digest: "SHA256", serializer: JSON, url_safe: true }
end

# Enable `transitional` mode for rolling deployment
Rails.application.message_verifiers.transitional = true

# `on_rotation` callback to log the usage of old verifiers
Rails.application.message_verifiers.on_rotation do
  puts "Old verifier used for verifying! Still in use!"
end
```

### Todos
- [x] Determine the Rails version in which the deprecated `signed_id_verifier_secret` will be removed.
- [x] Add tests for `railtie.rb` changes
- [x] Ensure CI passes
- [x] Update documentation
- [x] Update CHANGELOG.md file

jonathanhefner commented on activerecord/lib/active_record/signed_id.rb:
```diff
@@ -70,34 +76,12 @@ def find_signed(signed_id, purpose: nil, on_rotation: nil)
-      # The verifier instance that all signed ids are generated and verified from. By default, it'll be initialized
-      # with the class-level +signed_id_verifier_secret+, which within \Rails comes from the
-      # Rails.application.key_generator. By default, it's SHA256 for the digest and JSON for the serialization.
-      def signed_id_verifier
-        @signed_id_verifier ||= begin
-          secret = signed_id_verifier_secret
-          secret = secret.call if secret.respond_to?(:call)
-
-          if secret.nil?
-            raise ArgumentError, "You must set ActiveRecord::Base.signed_id_verifier_secret to use signed ids"
-          else
-            ActiveSupport::MessageVerifier.new secret, digest: "SHA256", serializer: JSON, url_safe: true
-          end
-        end
-      end
```
`signed_id_verifier_secret` is also a `class_attribute`, which is unfortunate.  That means existing apps could have, for example, set `Post.signed_id_verifier_secret` and `Comment.signed_id_verifier_secret` to different values, and `Post.signed_id_verifier` and `Comment.signed_id_verifier` each have to use those values, respectively.

I have not fully sketched this out, but I would try giving `ActiveRecord::Base` access to `Rails.application.message_verifiers`.  In other words, set something like `ActiveRecord::Base.message_verifiers` in an initializer, and use that instead of `MessageVerifier.new` in this method.

AliSepehri replied:
@jonathanhefner Good point. I intended to maintain the current configuration; however, doing so would prevent us from migrating to the default configuration.
I tried it here [https://github.com/rails/rails/pull/54474](https://github.com/rails/rails/pull/54474), please take a look.

jonathanhefner replied:
Is there a particular reason for a separate PR?  If not, I think you should add those changes to this PR so that we can consolidate the discussion.

AliSepehri replied:
I initially thought we might comeback to this solution if the new one doesn't prove to be better. In any case, I've incorporated the changes into this PR.

jonathanhefner commented on activerecord/lib/active_record/signed_id.rb:
```diff
@@ -52,7 +57,8 @@ module ClassMethods
       def find_signed(signed_id, purpose: nil, on_rotation: nil)
         raise UnknownPrimaryKey.new(self) if primary_key.nil?
 
-        if id = signed_id_verifier.verified(signed_id, purpose: combine_signed_id_purposes(purpose), on_rotation: on_rotation)
+        options = { on_rotation: on_rotation }.compact
+        if id = signed_id_verifier.verified(signed_id, purpose: combine_signed_id_purposes(purpose), **options)
```
Why this change?

AliSepehri replied:
The `Message::Rotator#read_message` method is defined like below:
```ruby
def read_message(message, on_rotation: @on_rotation, **options)
  ...
end
```
and when the caller sets `on_rotation` by `nil` it doesn't use `@on_rotation` value. So I'm using `compact` to filter out `nil` values.

jonathanhefner replied:
That is a bug introduced in #54373, isn't it?  In that case, it should be fixed in a separate PR.

AliSepehri replied:
Yes and no! Previously, it wasn't possible to set `@on_rotation` for `signed_id_verifier`. In other words, there was no way to write a test that would fail with the previous implementation. After using the centralized `message_verifiers` it can be considered as bug.

jonathanhefner replied:
But apps can specify a custom `signed_id_verifier`, no?  For example:

```ruby
Post.signed_id_verifier = ActiveSupport::MessageVerifier.new(..., on_rotation: callback)
```

AliSepehri replied:
Yeah that's right. I will try it and if it's the case I'll create a separate PR for it. 👍

AliSepehri replied:
@jonathanhefner You were right. Here is a separate PR for the fix that you mentioned, please review: https://github.com/rails/rails/pull/54523
🙏

jonathanhefner commented on activerecord/lib/active_record/signed_id.rb:
```diff
@@ -6,11 +6,10 @@ module SignedId
-      # :singleton-method:
+      mattr_accessor :signed_id_verifier
+
       # Set the secret used for the signed id verifier instance when using Active Record outside of \Rails.
-      # Within \Rails, this is automatically set using the \Rails application key generator.
-      class_attribute :signed_id_verifier_secret, instance_writer: false
+      mattr_accessor :signed_id_verifier_secret, instance_writer: false
```
`mattr_accessor` is not the same as using a singleton instance variable such as in the current definition of `signed_id_verifier`.  Changing `signed_id_verifier` or `signed_id_verifier_secret` to `mattr_accessor` will break subclassing behavior.

Did you try my suggestion from https://github.com/rails/rails/pull/54422#discussion_r1943343277?  Did you run into an issue?

AliSepehri replied:
You're absolutely right—it can break some configurations (I've also listed them in the "Impacts" section of the description). My intention was to follow your suggestion, so it seems I misunderstood you.  

From my understanding, `MessageVerifiers`/`RotationCoordinator` is designed in a way that requires a consistent configuration across all instances. However, there is a way to provide different configurations for rotation. Now, considering our requirements, we need to ensure:  

- The current default configuration for `signed_id_verifier`  
- The ability to have a separate `signed_id_verifier_secret` for each model  
- A unified initialization of message verifiers via `Rails.application.message_verifiers`  

Maybe something like:
```ruby
# initialization
ActiveRecord::Base.message_verifier = app.message_verifier("active_record/signed_id")

# signed_id.rb
message_verifier.rotate do |salt|
  next nil if salt != "active_record/signed_id"

  { secret_generator: signed_id_verifier_secret, digest: "SHA256", serializer: JSON, url_safe: true }
end
```

With this approach, each model will add a rotation, and we might be able to unify it before building them.  

I'm not sure—what do you think? Or did you mean something different?

jonathanhefner replied:
> Now, considering our requirements, we need to ensure:
>
> * The current default configuration for `signed_id_verifier`
> * The ability to have a separate `signed_id_verifier_secret` for each model
> * A unified initialization of message verifiers via `Rails.application.message_verifiers`

To clarify, the ultimate goal is for all verifiers to share the same configuration (derived from `Rails.application.message_verifiers`) so that it is straightforward to, for example, rotate `secret_key_base` or upgrade the signing algorithm.

So, taking a step back, I think we should deprecate `signed_id_verifier_secret`.  We cannot remove it (yet), but we can deprecate it.  We can make it such that if an app sets `signed_id_verifier_secret`, `signed_id_verifier` will be exactly the same as now (i.e. it will ignore `Rails.application.message_verifiers`).  Then, after a deprecation cycle, we can remove that code.

Next is the current default configuration for `signed_id_verifier`.  I've just added [`MessageVerifiers#prepend`](https://edgeapi.rubyonrails.org/classes/ActiveSupport/MessageVerifiers.html#method-i-prepend) in #54525.  My thinking is that we can add a `config.active_record.use_legacy_signed_id_verifier` setting that behaves like:

```ruby
# in an Active Record initializer

legacy_options = { ... }

if config.active_record.use_legacy_signed_id_verifier == :generate_and_verify
  app.message_verifiers.prepend { |salt| legacy_options if salt == "active_record/signed_id" }
elsif config.active_record.use_legacy_signed_id_verifier == :verify
  app.message_verifiers.rotate { |salt| legacy_options if salt == "active_record/signed_id" }
end
```

Though we will need to document that if `use_legacy_signed_id_verifier` is false, the signed IDs will not be URL-safe unless `Rails.application.message_verifiers` uses `url_safe: true`.

Separately from all of that, I think we should push for `url_safe` to default to true, for both `MessageVerifier` and `MessageEncryptor`.  But that will require its own migration path.

AliSepehri replied:
@jonathanhefner Sure! Thanks for your clear explanation. I've made the changes and updated the TODO list for the PR—please take a look. Using [MessageVerifiers#prepend](https://edgeapi.rubyonrails.org/classes/ActiveSupport/MessageVerifiers.html#method-i-prepend) was a great idea to resolve our issue!

jonathanhefner commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -166,6 +166,8 @@ class InvalidSignature < StandardError; end
     #   +config.active_support.use_message_serializer_for_metadata+.
     def initialize(secret, **options)
       raise ArgumentError, "Secret should not be nil." unless secret
+
+      options[:url_safe] = true if options[:url_safe].nil?
```
This would be a breaking change.  Why change it?

AliSepehri replied:
Due to [[this change](https://github.com/rails/rails/pull/49507)](https://github.com/rails/rails/pull/49507), `signed_id` needs to be generated in a URL-safe manner. Since this configuration applies to all message verifiers, I couldn't find a way other than changing this. However, I acknowledge that this could be a breaking change.

AliSepehri commented:
Oooops! 🫤 I just wanted to update my fork. I resynced wrong branch.

AliSepehri committed:
Use Rails.application.message_verifiers for signed IDs

Prior to this change, the primary way to configure signed ID verifiers
was to set `signed_id_verifier` on each model class:

  ```ruby
  Post.signed_id_verifier = ActiveSupport::MessageVerifier.new(...)
  Comment.signed_id_verifier = ActiveSupport::MessageVerifier.new(...)
  ```

And if the developer did not set `signed_id_verifier`, a verifier would
be instantiated with a secret derived from `secret_key_base` and the
following options:

  ```ruby
  { digest: "SHA256", serializer: JSON, url_safe: true }
  ```

Thus it was cumbersome to rotate configuration for all verifiers.

This change defines a new Rails config: `config.active_record.use_legacy_signed_id_verifier`.
The default value is `:generate_and_verify`, which preserves the
previous behavior.  However, when set to `:verify`, signed ID verifiers
will use configuration from `Rails.application.message_verifiers` to
generate and verify signed IDs, but will also verify signed IDs using
the older configuration.

To avoid complication, the new behavior only applies when `signed_id_verifier_secret`
is not set on a model class, and `signed_id_verifier_secret` is now
deprecated.  Instead of setting `signed_id_verifier_secret`, developers
can set `signed_id_verifier`:

  ```ruby
  # BEFORE
  Post.signed_id_verifier_secret = "my secret"

  # AFTER
  Post.signed_id_verifier = ActiveSupport::MessageVerifier.new("my secret", digest: "SHA256", serializer: JSON, url_safe: true)
  ```

To ease migration, `signed_id_verifier` has also been changed to behave as a
`class_attribute` (i.e. inheritable), but _only when `signed_id_verifier_secret`
is not set_:

  ```ruby
  # BEFORE
  ActiveRecord::Base.signed_id_verifier = ActiveSupport::MessageVerifier.new(...)
  Post.signed_id_verifier == ActiveRecord::Base.signed_id_verifier # => false

  # AFTER
  ActiveRecord::Base.signed_id_verifier = ActiveSupport::MessageVerifier.new(...)
  Post.signed_id_verifier == ActiveRecord::Base.signed_id_verifier # => true

  Post.signed_id_verifier_secret = "my secret" # => deprecation warning
  Post.signed_id_verifier == ActiveRecord::Base.signed_id_verifier # => false
  ```

Co-authored-by: Jonathan Hefner <jonathan@hefner.pro>

AliSepehri commented on activerecord/lib/active_record/signed_id.rb:
```diff
@@ -8,9 +8,20 @@ module SignedId
+      # <tt>:generate_and_verify</tt> – Generates signed IDs using the legacy format while verifying both the new and legacy formats.
+      # <tt>:verify</tt> (default) – Generates signed IDs using the new format while verifying both the new and legacy formats.
+      # Note: The new format is not URL-safe unless <tt>Rails.application.message_verifiers</tt> is configured with <tt>url_safe: true</tt>.
+      mattr_accessor :use_legacy_signed_id_verifier, default: :verify
```
❓ Should use `:generate_and_verify` as the default value until `message_verifiers` is updated to use `url_safe: true`?

AliSepehri replied:
I tried to find the right place to define this config attribute. I also updated the `configuring.md`.

AliSepehri commented on activerecord/lib/active_record/railtie.rb:
```diff
@@ -319,9 +319,26 @@ class Railtie < Rails::Railtie # :nodoc:
-    initializer "active_record.set_signed_id_verifier_secret" do
-      ActiveSupport.on_load(:active_record) do
-        self.signed_id_verifier_secret ||= -> { Rails.application.key_generator.generate_key("active_record/signed_id") }
+    initializer "active_record.set_signed_id_verifier" do
+      config.after_initialize do |app|
+        ActiveSupport.on_load(:active_record) do
+          secret = self.signed_id_verifier_secret
+          secret_generator = if secret
+            ActiveRecord.deprecator.warn(<<~MSG)
+              ActiveRecord::Base.signed_id_verifier_secret is deprecated and will be removed in Rails ?.?. The secret will be derived from 'secret_key_base' instead.
+            MSG
+            secret.respond_to?(:call) ? secret : -> (_) { secret }
+          end
+          legacy_options = { secret_generator: secret_generator, digest: "SHA256", serializer: JSON, url_safe: true }.compact
+
+          if self.use_legacy_signed_id_verifier == :generate_and_verify
+            app.message_verifiers.prepend { |salt| legacy_options if salt == "active_record/signed_id" }
+          elsif self.use_legacy_signed_id_verifier == :verify
+            app.message_verifiers.rotate { |salt| legacy_options if salt == "active_record/signed_id" }
+          end
+
+          self.signed_id_verifier ||= app.message_verifier("active_record/signed_id")
+        end
```
❓ Where is the best place to test this functionality?

jonathanhefner replied:
[`railties/test/application/configuration_test.rb`](https://github.com/rails/rails/blob/main/railties/test/application/configuration_test.rb) is a good place to test configuration from initializers.

AliSepehri replied:
We had one to test the default value and added one test for each mode.

AliSepehri commented on activerecord/lib/active_record/railtie.rb:
```diff
@@ -319,9 +319,26 @@ class Railtie < Rails::Railtie # :nodoc:
+          secret = self.signed_id_verifier_secret
+          secret_generator = if secret
+            ActiveRecord.deprecator.warn(<<~MSG)
+              ActiveRecord::Base.signed_id_verifier_secret is deprecated and will be removed in Rails ?.?. The secret will be derived from 'secret_key_base' instead.
```
❓ What is the Rails version that we can remove `signed_id_verifier_secret`?

AliSepehri replied:
I moved deprecation message into `ActiveRecord::SignedId`. I believe the rest of the logic should stay here in `railtie.rb ` file. wdyt?

AliSepehri replied:
> signed_id_verifier_secret is also a class_attribute, which is unfortunate. That means existing apps could have, for example, set Post.signed_id_verifier_secret and Comment.signed_id_verifier_secret to different values, and Post.signed_id_verifier and Comment.signed_id_verifier each have to use those values, respectively.

We still don't support this one. I'm not sure what should we do in this case? It's very complicated to handle.

AliSepehri replied:
I think I have an idea on how to approach this.  

If a `secret` is set for `ActiveRecord::Base`, we can use `message_verifiers` in `:generate_and_verify` mode. However, models that define their own `secret` should use a custom `signed_id_verifier`.  

In other words, there should be a global verifier initialized by `MessageVerifiers`, while models with a different `secret` can have their own custom verifiers.

jonathanhefner replied:
Yes, I touched on that a bit in https://github.com/rails/rails/pull/54422#discussion_r1943343277 :smiley::

> I have not fully sketched this out, but I would try giving `ActiveRecord::Base` access to `Rails.application.message_verifiers`. In other words, set something like `ActiveRecord::Base.message_verifiers` in an initializer, and use that instead of `MessageVerifier.new` in this method.

So instead of setting `ActiveRecord::Base.global_signed_id_verifier = app.message_verifier("active_record/signed_id")` in an initializer, we would set `ActiveRecord::Base.message_verifiers = app.message_verifiers`.

Then, in the `signed_id_verifier` method, we would return `ActiveRecord::Base.message_verifiers["active_record/signed_id"]` if `signed_id_verifier_secret` was not set.

AliSepehri replied:
Ah yeah! 😃
I'll push the required changes for it.

AliSepehri replied:
I'm using `class_attribute` for the `_signed_id_verifier` attribute, which introduces a slight change from the previous approach. Previously, we used `@signed_id_verifier`.  

I assume that to use `ActiveRecord` outside of Rails, `signed_id_verifier` should be explicitly set instead of relying on `signed_id_verifier_secret`. Setting `signed_id_verifier` for all models didn’t seem like a practical solution, so I opted to use `class_attribute` instead.

jonathanhefner commented on guides/source/configuring.md:
```diff
@@ -1684,6 +1684,14 @@ required:
 
+#### `config.active_record.use_legacy_signed_id_verifier`
+
+This option allows for a smooth transition during the deprecation of the legacy `signed_id_verifier`. It supports the following modes:
+- `:generate_and_verify` (default) – Generates signed IDs using the legacy format while verifying both the new and legacy formats.
+- `:verify` – Generates signed IDs using the new format while verifying both the new and legacy formats.
```
We need to include the "false" option here too.

AliSepehri replied:
Yes, totally agree. I was also thinking about it. 👍

jonathanhefner reviewed (commented):
I've left some new comments, but I'm also pushing a commit to address them.  I did not work on any of the unit tests though.  Take a look, and let me know if you see any issues.

AliSepehri commented on activerecord/lib/active_record.rb:
```diff
@@ -469,6 +469,16 @@ def self.permanent_connection_checkout=(value)
+
+  # Note: The new format is not URL-safe unless +Rails.application.message_verifiers+ is configured with +url_safe: true+.
+  singleton_class.attr_accessor :use_legacy_signed_id_verifier
+  self.use_legacy_signed_id_verifier = :generate_and_verify
```
I see, yeah exactly I added that to fix the error.

AliSepehri commented on activerecord/lib/active_record/core.rb:
```diff
@@ -102,6 +102,8 @@ def self.configurations
 
       class_attribute :shard_selector, instance_accessor: false, default: nil
 
+      class_attribute :message_verifiers, instance_accessor: false, instance_predicate: false, default: nil
```
> `class_attribute` means that each model class can potentially have its own instance of `message_verifiers`.

Exactly, I wasn't sure what is the most correct way to define this variable. 👍 

> maybe this should actually be `ActiveRecord.message_verifiers` instead

Good idea, let's do it that way.

AliSepehri commented on activerecord/lib/active_record/railtie.rb:
```diff
@@ -319,9 +320,31 @@ class Railtie < Rails::Railtie # :nodoc:
+        ActiveSupport.on_load(:active_record) do
+          self.message_verifiers = app.message_verifiers
+        end
       end
```
Nice! 👌

AliSepehri commented on activerecord/lib/active_record/signed_id.rb:
```diff
@@ -11,6 +11,21 @@ module SignedId
       # Within \Rails, this is automatically set using the \Rails application key generator.
       class_attribute :signed_id_verifier_secret, instance_writer: false
+
+      # :nodoc:
```
I didn't know! 👍

AliSepehri commented on activerecord/lib/active_record/signed_id.rb:
```diff
@@ -11,6 +11,21 @@ module SignedId
+        # :nodoc:
+        def signed_id_verifier_secret=(secret)
+          ActiveRecord.deprecator.warn(<<~MSG)
+            ActiveRecord::Base.signed_id_verifier_secret is deprecated and will be removed in the future. The secret will be derived from 'secret_key_base' instead.
```
👍

AliSepehri commented on activerecord/lib/active_record/railtie.rb:
```diff
@@ -319,9 +320,31 @@ class Railtie < Rails::Railtie # :nodoc:
+          app.message_verifiers.rotate { |salt| legacy_options if salt == "active_record/signed_id" }
+        else
+          raise ArgumentError, "Unknown value for ActiveRecord::Base.use_legacy_signed_id_verifier: #{app.config.active_record.use_legacy_signed_id_verifier}"
+        end
```
Got it! Thanks for the change.

AliSepehri commented:
> I've left some new comments, but I'm also pushing a commit to address them. I did not work on any of the unit tests though. Take a look, and let me know if you see any issues.

@jonathanhefner Thanks for the changes. I reviewed all and I didn't see any issue. 👌 I just pushed some unit tests for railtie initializer. What's remaining you think? Do we need to update CHANGELOG.md file?

AliSepehri commented:
@jonathanhefner Do you think, we're ready with this PR?

AliSepehri commented:
@jonathanhefner I added a bunch of tests and tried to cover the scenarios you mentioned. Initially, it wasn’t clear to me where to add tests for the new code, but thanks for your suggestions!

AliSepehri commented on activerecord/test/cases/signed_id_test.rb:
```diff
@@ -223,6 +227,54 @@ def set_signed_id
+    assert_not_equal account_signed_id_with_message_verifiers, account_signed_id_with_secret
+  ensure
+    ActiveRecord.deprecator.silence do
+      Account.signed_id_verifier_secret = nil
```
@jonathanhefner Since `signed_id_verifier_secret uses` a `class_attribute`, even reassigning it to `nil` cannot completely clean up the side effects. For example:
```ruby
Account.signed_id_verifier_secret = "account secret"
Account.signed_id_verifier_secret = nil
ActiveRecord::Base.signed_id_verifier_secret = "secret"

puts Account.signed_id_verifier_secret # => nil
```
Do you know any solution or we can ignore it for now?

jonathanhefner commented on activerecord/test/cases/signed_id_test.rb:
```diff
@@ -223,6 +227,54 @@ def set_signed_id
+    end
+  end
+
+  test "model-specific signed_id_verifier_secret does not affect other models" do
+    toy_signed_id = @toy.signed_id
+
+    ActiveRecord.deprecator.silence do
+      Account.signed_id_verifier_secret = "account secret"
+    end
+
+    assert_equal @toy.signed_id, toy_signed_id
+  ensure
+    ActiveRecord.deprecator.silence do
+      Account.signed_id_verifier_secret = nil
+      Account.instance_variable_set :@signed_id_verifier, nil
+    end
+  end
```
Hmm, I don't this test is necessary.  I think it is sufficient to test that `ActiveRecord::Base.signed_id_verifier` is not affected (which we do in another test).  Is there a case I am missing?

AliSepehri replied:
No you're right. Checking `ActiveRecord::Base.signed_id_verifier` can be enough.

jonathanhefner reviewed (commented):
I've made another round of comments, and I pushed a commit to address them.  Take a look.

I also want to tweak the CHANGELOG entry to elaborate on the way things currently work and the motivation for the change, but I will do that later.

AliSepehri commented on railties/test/application/configuration_test.rb:
```diff
@@ -2943,21 +2943,13 @@ def index
-        ActiveSupport::MessageVerifier.new(app.key_generator("secret").generate_key("active_record/signed_id"))
-          .generate(1, purpose: "account")
-      assert signed_id_verifier.verify(signed_account_id_one_with_new_options, purpose: "account")
+      assert_equal "message", legacy_verifier.verify(signed_id_verifier.generate("message"))
+      assert_equal "message", signed_id_verifier.verify(legacy_verifier.generate("message"))
```
This is much cleaner. I chose to compare the signed ID itself because [this](https://github.com/rails/rails/commit/f643919993f0de126f2d99604972bb0ba58a6296) change allows both URL-safe and non-URL-safe versions. But I agree with your implementation and we could ignore it.

AliSepehri commented on activerecord/lib/active_record/signed_id.rb:
```diff
@@ -77,28 +93,38 @@ def find_signed!(signed_id, purpose: nil, on_rotation: nil)
+
+          if ActiveRecord.message_verifiers.nil?
+            raise "You must set ActiveRecord.message_verifiers to use signed ids"
+          end
```
right, I had missed it. 👍

AliSepehri commented on activerecord/test/cases/signed_id_test.rb:
```diff
@@ -223,6 +227,54 @@ def set_signed_id
+    account_signed_id_with_secret = @account.signed_id
+    assert_equal @account, Account.find_signed(account_signed_id_with_secret)
+
+    assert_not_equal account_signed_id_with_message_verifiers, account_signed_id_with_secret
```
right! 👍 Thanks for the changes.

AliSepehri commented:
> I've made another round of comments, and I pushed a commit to address them. Take a look.
> 
> I also want to tweak the CHANGELOG entry to elaborate on the way things currently work and the motivation for the change, but I will do that later.

@jonathanhefner Thanks for the comments and changes. I reviewed all and they look good to me! 👍

AliSepehri commented:
@jonathanhefner Do you want me to prepare a draft for CHANGELOG?

jonathanhefner commented:
@AliSepehri  I pushed a commit to revise the CHANGELOG entry.  I wanted to clarify the "before" and "after" behaviors, and to communicate that the developer doesn't necessarily need to change anything unless they are setting `signed_id_verifier_secret` on a model (since that is now deprecated).  I also rebased and squashed the commits.

Have a look at the CHANGELOG entry, and let me know what you think.

AliSepehri commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,65 @@
+    older configuration.
+
+    To avoid complication, the new behavior only applies when `signed_id_verifier_secret`
+    is not set on a model class, and `signed_id_verifier_secret` is now
```
Since it's also possible to set `signed_id_verifier_secret` for `ActiveRecord::Base`, perhaps we could mention "for ActiveRecord or a specific model class". wdyt?

jonathanhefner replied:
Sounds good.  I changed it to:

  ```diff
  --- a/activerecord/CHANGELOG.md
  +++ b/activerecord/CHANGELOG.md
  @@ -27,9 +27,10 @@
       older configuration.

       To avoid complication, the new behavior only applies when `signed_id_verifier_secret`
  -    is not set on a model class, and `signed_id_verifier_secret` is now
  -    deprecated. If you are currently setting `signed_id_verifier_secret` on a
  -    model class, set `signed_id_verifier` instead:
  +    is not set on a model class or any of its ancestors. Additionally,
  +    `signed_id_verifier_secret` is now deprecated. If you are currently setting
  +    `signed_id_verifier_secret` on a model class, you can set `signed_id_verifier`
  +    instead:

         ```ruby
         # BEFORE
  ```

since it is affected by any ancestor class (e.g. `ApplicationRecord`).

AliSepehri replied:
👍🏽

AliSepehri commented on activerecord/CHANGELOG.md:
````diff
@@ -1,3 +1,65 @@
+      # BEFORE
+      Post.signed_id_verifier_secret = "my secret"
+
+      # AFTER
+      Post.signed_id_verifier = ActiveSupport::MessageVerifier.new("my secret", digest: "SHA256", serializer: JSON, url_safe: true)
+      ```
````
And perhaps we should also add instructions on how they can eventually migrate smoothly to `message_verifiers`?

Something like the code below to an initializer:
```ruby
legacy_options = { secret_generator: Proc.new { "secret" }, digest: "SHA256", serializer: JSON, url_safe: true }
ActiveRecord.message_verifiers.rotate { |salt| legacy_options if salt == "active_record/signed_id" }
```

jonathanhefner replied:
The point of this change is to eventually achieve a unified configuration, so I do not want to encourage developers to do something like `message_verifiers.rotate { |salt| options if salt == "active_record/signed_id" }`.  If they already have model-specific verifiers, they should use `rotate` on those verifiers instead (and eventually remove them).

I added the following to further explain:

  ```diff
  --- a/activerecord/CHANGELOG.md
  +++ b/activerecord/CHANGELOG.md
  @@ -56,6 +57,23 @@
         Post.signed_id_verifier == ActiveRecord::Base.signed_id_verifier # => false
         ```

  +    Note, however, that it is recommended to eventually migrate from
  +    model-specific verifiers to a unified configuration managed by
  +    `Rails.application.message_verifiers`. `ActiveSupport::MessageVerifier#rotate`
  +    can facilitate that transition. For example:
  +
  +      ```ruby
  +      # BEFORE
  +      # Generate and verify signed Post IDs using Post-specific configuration
  +      Post.signed_id_verifier = ActiveSupport::MessageVerifier.new("post secret", ...)
  +
  +      # AFTER
  +      # Generate and verify signed Post IDs using the unified configuration
  +      Post.signed_id_verifier = Post.signed_id_verifier.dup
  +      # Fall back to Post-specific configuration when verifying signed IDs
  +      Post.signed_id_verifier.rotate("post secret", ...)
  +      ```
  +
  ```

(I also submitted #54873 so that `signed_id_verifier.dup` behaves as expected.)

AliSepehri replied:
Sounds great! 👌🏽

AliSepehri reviewed (commented):
@jonathanhefner Thanks for updating the CHANGELOG. It's much clear now. I just added two non-blocking comments.

AliSepehri commented:
@jonathanhefner We’re ready now! Shall we?

jonathanhefner commented:
Thank you, @AliSepehri! :shipit:

AliSepehri commented:
And thank you, @jonathanhefner! 👍
